File: backend/app/routers/patients.py
"""[MASK-*][SEARCH-LOG-*][ROLE-READ] patients 라우터 — 마스킹 목록 · 상세 · 번호 펼치기.

⚠️ 코디 배선 필요: main.py에 `app.include_router(patients.router)` 등록해야 노출된다.
   이 태스크는 main.py를 손대지 않는다(공용 파일).

접근 범위: 목록·상세·번호 펼치기 모두 접수직원·관리자만(patients RLS의
receptionist_admin_can_read_patients와 일치). 의사의 조회 범위는 자기 예약이라
환자 목록 전체 창구는 열지 않는다(ROLE-DOC-02).
"""
from datetime import date
from uuid import UUID

from fastapi import APIRouter, Depends
from pydantic import BaseModel

from app.core.dto import patient_row_dto
from app.core.security import StaffContext, require_role
from app.services import patient_service, staff_phone_change_service

router = APIRouter(prefix="/patients", tags=["patients"])


class RegisterPatientRequest(BaseModel):
    # [SHELL-DOOR-03] 등록 문의 신원 폼 — 이름·성별·생년월일·전화. 검색칸·담당의사는 폼에서 뺀다.
    name: str
    gender: str
    birth_date: date
    phone: str


class RegisterPatientResponse(BaseModel):
    patient_id: UUID


class DuplicateCheckResponse(BaseModel):
    # [SHELL-DOOR-03] "혹시 이분?" — 후보 id + 표시값, 없으면 전부 null. ⛔ 막지 않는다(등록 게이트가 아니다).
    # ⭐ 이름은 실명이다(요구사항 :81은 전화·생년월일만 가리라고 한다). 원본 birth_date·phone은 안 담는다.
    patient_id: UUID | None = None
    name: str | None = None
    masked_birth_date: str | None = None


class FamilyLinkRequest(BaseModel):
    family_patient_id: UUID
    relation: str
    method: str


class FamilyUnlinkRequest(BaseModel):
    reason: str


class PhoneChangeRequestBody(BaseModel):
    # [PTDET-ACTION-02] 새 전화번호. 서버가 이 번호로 인증번호를 보낸다(㉯ 소유 증명).
    new_phone: str


class PhoneChangeConfirmBody(BaseModel):
    # [PTDET-ACTION-02] 확인은 (patient_id, new_phone)로 요청을 찾는다 — 화면이 request_id를 들지 않는다.
    new_phone: str
    code: str


@router.get("")
async def list_patients(
    q: str | None = None,
    cursor: str | None = None,
    staff: StaffContext = Depends(require_role("receptionist", "admin", "doctor")),
) -> dict:
    """[MASK-SRV-01][SEARCH-*] 검색 결과(마스킹) + 다음 페이지 커서 + 검색 기록.

    ⭐ 마스킹 경계는 여기다 — 서비스는 정렬을 위해 원본을 담은 줄을 주고, HTTP로 나가는
       이 지점에서 patient_row_dto가 masked_* 로만 옮긴다(원본 키는 응답에 없다). 줄마다
       matched(왜 걸렸는지)·오늘 상태·오늘 예약 시각을 함께 실어 24b가 그대로 소비한다.
    """
    page = await patient_service.search_patients(q, staff, cursor=cursor)
    return {
        "rows": [
            patient_row_dto(
                patient_id=row["id"],
                name=row["name"],
                phone=row["phone"],
                birth_date=row["birth_date"],
                gender=row["gender"],
                matched=row["matched"],
                today_status=row["today_status"],
                today_appointment_time=row["today_appointment_time"],
                today_department_name=row["today_department_name"],
                today_doctor_name=row["today_doctor_name"],
            )
            for row in page.rows
        ],
        "next_cursor": page.next_cursor,
        "has_more": page.has_more,
    }


@router.post("", status_code=201, response_model=RegisterPatientResponse)
async def register_patient(
    body: RegisterPatientRequest,
    staff: StaffContext = Depends(require_role("receptionist", "admin")),
) -> RegisterPatientResponse:
    """[SHELL-DOOR-03][ROLE-READ] 신원 폼으로 환자를 등록한다 — 접수직원·관리자만.

    소프트 중복은 등록을 막지 않는다(개인정보 열거 방지, 「막다른 길 금지」). "혹시 이분?"은
    duplicate-check가 화면에 힌트를 줄 뿐, 이 창구는 겹침 여부와 무관하게 등록을 진행시킨다.
    """
    patient_id = await patient_service.register_patient(
        name=body.name,
        birth_date=body.birth_date,
        gender=body.gender,
        phone=body.phone,
        staff=staff,
    )
    return RegisterPatientResponse(patient_id=patient_id)


@router.get("/duplicate-check", response_model=DuplicateCheckResponse)
async def duplicate_check(
    phone: str,
    birth_date: date,
    staff: StaffContext = Depends(require_role("receptionist", "admin")),
) -> DuplicateCheckResponse:
    """[SHELL-DOOR-03] 소프트 중복 — 전화·생일이 강하게 겹치는 기존 기록의 id(또는 null).

    ⚠️ 라우트 순서: 이 경로는 반드시 `/{patient_id}`(UUID)보다 **먼저** 선언해야 한다.
       뒤에 두면 FastAPI가 "duplicate-check"를 patient_id로 파싱하려다 422를 낸다.
    ⛔ 막지 않는다 — 후보를 알려줄 뿐 등록을 거부하지 않는다(개인정보 열거 방지·막다른 길 금지).
    """
    row = await patient_service.find_by_phone_and_birthdate(phone, birth_date, staff)
    if row is None:
        return DuplicateCheckResponse()
    # 마스킹 경계는 여기다 — 서비스가 준 원본 줄을 화이트리스트 DTO로 옮겨 가린 값만 남긴다.
    dto = patient_row_dto(
        patient_id=row["id"], name=row["name"], birth_date=row["birth_date"]
    )
    return DuplicateCheckResponse(**dto)


@router.get("/{patient_id}")
async def get_patient(
    patient_id: UUID,
    staff: StaffContext = Depends(require_role("receptionist", "admin", "doctor")),
) -> dict:
    """[MASK-DETAIL-01] 상세(전체) + 진입 기록.

    [SHELL-NAV-03][ROLE-DOC-02] 의사도 사이드바 「환자 검색」→ 상세로 온다. RLS
    `doctor_can_read_scoped_patients`가 본인 담당 환자로 스코프하고, 아니면 404(열거 안전).
    """
    return await patient_service.get_patient_detail(patient_id, staff)


@router.get("/{patient_id}/contact")
async def reveal_contact(
    patient_id: UUID,
    staff: StaffContext = Depends(require_role("receptionist", "admin")),
) -> dict:
    """[MASK-VIEW-01·02·03] 번호 펼치기 창구(갭 #35) + 열람 기록."""
    return await patient_service.reveal_contact(patient_id, staff)


@router.post("/{patient_id}/family")
async def link_family(
    patient_id: UUID,
    body: FamilyLinkRequest,
    staff: StaffContext = Depends(require_role("receptionist", "admin")),
) -> dict:
    """[R5-01][PTDET-FAMILY-04·05][ROLE-DOC-02] 가족 연결 저장 창구 — 접수·관리자만.

    의사는 이 창구를 열 수 없다(가족 연결은 접수·관리자의 일, 요구사항 3.5·4.2).
    """
    link_id = await patient_service.link_family_member(
        patient_id, body.family_patient_id, body.relation, body.method, staff
    )
    return {"id": link_id}


@router.delete("/{patient_id}/family/{member_id}")
async def unlink_family(
    patient_id: UUID,
    member_id: UUID,
    body: FamilyUnlinkRequest,
    staff: StaffContext = Depends(require_role("receptionist", "admin")),
) -> dict:
    """[R5-02][ROLE-DOC-02] 가족 연결 해제 — 접수·관리자만. 사유·실행자를 남긴다."""
    await patient_service.unlink_family_member(patient_id, member_id, body.reason, staff)
    return {"ok": True}


@router.post("/{patient_id}/phone-change/request")
async def phone_change_request(
    patient_id: UUID,
    body: PhoneChangeRequestBody,
    staff: StaffContext = Depends(require_role("receptionist", "admin")),
) -> dict:
    """[PTDET-ACTION-02][갭 #19·결정 #4 ㉯] 새 번호로 인증번호를 보낸다 — 접수·관리자만.

    의사는 이 창구를 열 수 없다(전화번호 변경은 접수·관리자의 일, ROLE-DOC-02).
    새 번호 OTP 소유 증명이 있어야 바꾼다 — 직접 저장(㉮)은 계정 탈취로 기각(결정 #4).
    """
    await staff_phone_change_service.request_phone_change(staff, patient_id, body.new_phone)
    return {"ok": True}


@router.post("/{patient_id}/phone-change/confirm")
async def phone_change_confirm(
    patient_id: UUID,
    body: PhoneChangeConfirmBody,
    staff: StaffContext = Depends(require_role("receptionist", "admin")),
) -> dict:
    """[PTDET-ACTION-02·03][결정 #4 ⓑ] 인증번호가 맞으면 patients.phone + Auth를 함께 바꾼다.

    실패하면 기존 번호가 산다(ACTION-03) — 성공 응답을 흉내 내지 않는다.
    """
    await staff_phone_change_service.confirm_phone_change(
        staff, patient_id, body.new_phone, body.code)
    return {"ok": True}

File: frontend/src/api/patients.ts
import { apiFetch } from './httpClient'

// 환자 도메인 얇은 클라이언트.
// ⚠️ 백엔드 라우터(backend/app/routers/patients.py)는 Task 6에서 처음 생긴다 —
//    경로는 Task 6 플랜의 계약(`GET /patients?q=` · `GET /patients/{id}` · `GET /patients/{id}/contact`)에 맞춘다.
//    전화번호 펼치기는 열람 기록이 남는 별도 창구(`/contact`)라, 목록·상세와 나눠 부른다(MASK-*·SEARCH-LOG-*).

// ⭐ 전역 환자 검색의 단일 창구(SEARCH-BOX-03) — /patients 화면과 워크인·전화예약 패널이 같은 걸 부른다.
//    24a(백엔드)가 못박은 계약을 그대로 소비한다: 한 칸(q) + 커서(cursor), 응답은 마스킹된 줄 + 커서.
//    ⛔ 원본(전화·생일 전체)은 응답에 오지 않는다 — 서버가 masked_* 로만 내려준다(MASK-SRV-01).

/** 왜 걸렸는지(SEARCH-WHY-01·03) — 조각별로 이름·전화·생일 중 무엇이 맞았나. */
export type SearchMatch = 'name' | 'phone' | 'birth'

/** 오늘 상태(SEARCH-ACT-*) — /queue와 같은 순간의 값을 서버가 준다. 화면이 다시 계산하지 않는다. */
export type SearchTodayStatus = 'booked' | 'arrived' | 'done' | null

/** 검색 결과 한 줄 — 마스킹된 표시값만. 24a 계약(patient_row_dto + matched·오늘상태). */
export interface SearchPatientRow {
  patient_id: string
  name: string
  masked_phone: string
  masked_birth_date: string
  gender: string | null
  matched: SearchMatch[]
  today_status: SearchTodayStatus
  /** "HH:MM" — 오늘 예약이 있을 때만. 맨 위에 있는 이유를 그 줄에 싣는다(SEARCH-ORDER-06). */
  today_appointment_time: string | null
  /** 오늘 예약의 진료과·의사 표시명 — 예약이 있을 때만(접수에 필요한 정보, SEARCH-ORDER-06). */
  today_department_name: string | null
  today_doctor_name: string | null
}

/** 커서로 이어받는 한 페이지(SEARCH-RESULT-02·03) — 20건·안정 동점키는 서버(paginate)가 소유. */
export interface SearchPatientsPage {
  rows: SearchPatientRow[]
  next_cursor: string | null
  has_more: boolean
}

export function searchPatients(query: string, cursor?: string | null) {
  const params = new URLSearchParams()
  if (query) params.set('q', query)
  if (cursor) params.set('cursor', cursor)
  const qs = params.toString()
  return apiFetch<SearchPatientsPage>(`/patients${qs ? `?${qs}` : ''}`)
}

export function getPatient(patientId: string) {
  return apiFetch<Record<string, unknown>>(`/patients/${patientId}`)
}

// 마스킹된 번호를 펼친다 — 서버가 열람 기록을 남기는 창구(갭 #35).
export function revealContact(patientId: string) {
  return apiFetch<Record<string, unknown>>(`/patients/${patientId}/contact`)
}

// ── 환자 상세 하위 이력 (PTDET-*) ────────────────────────────────────────────
// 백엔드 계약: backend/app/routers/dashboard.py · patient_history_service.
// ⭐ 섹션마다 독립적으로 부른다(PTDET-LOAD-02) — 하나의 Promise.all로 묶지 않는다.
//    한 섹션이 403이어도 나머지가 무너지지 않게, 소비 화면이 쿼리를 나눠 건다.

/** 상세 헤더 데이터 — 목록이 아니므로 원본(전체)로 온다(MASK-DETAIL-01). */
export interface PatientDetail {
  id: string
  name: string
  birth_date: string
  gender: string | null
  phone: string | null
  /** [SEND-DEAD-01] 이 번호로 문자가 가지 않음(00014 patients.sms_dead). */
  sms_dead?: boolean | null
  /** [SEND-DEAD-01] 문자 죽음 확인 시각(00014 patients.sms_dead_checked_at). */
  sms_dead_checked_at?: string | null
}

/** 마스킹된 이력 한 행 — 방문·진료기록 공통(patient_row_dto). */
export interface PatientHistoryRow {
  patient_id: string
  name?: string
  masked_birth_date?: string
  masked_phone?: string
  id: string
  occurred_at: string
  status?: string
  diagnosis?: string | null
  is_completed?: boolean
  relation?: string
  // 진료과·담당 의사 — 서버 방문 응답이 아직 담지 않는다(BLOCKED). 오면 그대로 그린다.
  department_name?: string | null
  doctor_name?: string | null
  // [PTDET-QNR-03 A안] 이 예약의 사전문진 제출 시각 — 있으면 작성완료, null/없으면 미작성.
  //   '작성 여부'만 온다(답변 내용은 담당 의사 전용, :420·#14). 서버: questionnaire_submitted_at_for(00076).
  questionnaire_submitted_at?: string | null
}

/** 공용 커서 페이지(core.pagination) — 방문·진료기록이 같은 형태로 온다. */
export interface HistoryPage {
  rows: PatientHistoryRow[]
  next_cursor: string | null
  has_more: boolean
}

export function getPatientDetail(patientId: string) {
  return apiFetch<PatientDetail>(`/patients/${patientId}`)
}

export function getPatientVisits(patientId: string, cursor?: string | null) {
  const q = cursor ? `?cursor=${encodeURIComponent(cursor)}` : ''
  return apiFetch<HistoryPage>(`/patients/${patientId}/visits${q}`)
}

export function getPatientMedicalRecords(patientId: string, cursor?: string | null) {
  const q = cursor ? `?cursor=${encodeURIComponent(cursor)}` : ''
  return apiFetch<HistoryPage>(`/patients/${patientId}/medical-records${q}`)
}

export function getPatientFamily(patientId: string) {
  return apiFetch<PatientHistoryRow[]>(`/patients/${patientId}/family`)
}

/** 내부 메모 한 줄 — 직원이 쓴 글이라 마스킹을 거치지 않는다(PTDET-NOTE-01). */
export interface PatientNote {
  id: string
  content: string
  created_at: string
  staff_name: string
}

export function getPatientNotes(patientId: string) {
  return apiFetch<PatientNote[]>(`/patients/${patientId}/notes`)
}

export function addPatientNote(patientId: string, content: string) {
  return apiFetch<{ id: string }>(`/patients/${patientId}/notes`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ content }),
  })
}

/** 사전문진 응답 — 담당 의사만 answers를 받는다(PTDET-QNR-03·RLS). */
export interface Questionnaire {
  appointment_id: string
  template_id: string
  answers: Record<string, unknown>
  submitted_at: string | null
}

export function getQuestionnaire(appointmentId: string) {
  return apiFetch<{ questionnaire: Questionnaire | null }>(
    `/appointments/${appointmentId}/questionnaire`,
  )
}

/** [PTDET-ACTION-02][갭 #19·결정 #4 ㉯] 새 번호로 인증번호를 보낸다 — 그 번호에 닿는 사람만 바꾼다. */
export function requestPhoneChange(patientId: string, newPhone: string) {
  return apiFetch<{ ok: boolean }>(`/patients/${patientId}/phone-change/request`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ new_phone: newPhone }),
  })
}

/** [PTDET-ACTION-02·03][결정 #4 ⓑ] 인증번호가 맞아야 번호가 바뀐다 — 실패하면 기존 번호가 산다. */
export function confirmPhoneChange(patientId: string, newPhone: string, code: string) {
  return apiFetch<{ ok: boolean }>(`/patients/${patientId}/phone-change/confirm`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ new_phone: newPhone, code }),
  })
}

/** 예외 진입 자격을 서버가 다시 판정한다(PTDET-FAMILY-04·05) — 화면이 정하지 않는다. */
export function verifyFamilyEligibility(patientId: string, memberId: string) {
  return apiFetch<{ allowed: boolean; message: string }>(
    `/patients/${patientId}/family/${memberId}/verify-eligibility`,
    { method: 'POST' },
  )
}
